Add `must-use-output` attribute (#3773)
rust-lang/rfcs: text/3773-must-use-output.md

Note also that some functions may be useless to call even when they take in truly immutable (`Freeze`) references. For instance `clone` is wasteful if the original is not accessed.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

You may use the `#[must_use_output]` attribute in front of any function argument with a reference type (`&T` or `&mut T`) to indicate to the caller that calling the function is useless or dangerous if the referenced value is not later accessed, e.g. reading it to check status. If the caller doesn't access the value later in the function a warning will be emitted.

As an example, this code:

```rust
fn main() {
    let mut vec = Vec::new();
    vec.push(42);
}
```

will emit a warning saying "`vec` is not used after the call to `push`" because `Vec::push` has `self` marked with this attribute:

```rust
impl<T> Vec<T> {
    pub fn push(#[must_use_output] &mut self, item: T) { /* ... */ }
}
```

> This is conceptually similar to `#[must_use]` on the function referring to the output value and helps you (or downstream consumers) to catch likely mistakes.

Reviewer: ```suggestion
This is conceptually similar to `#[must_use]` on the function, but for an output parameter rather than the return value. Like `#[must_use]`, it helps you (or downstream consumers) to catch likely mistakes.
```

A common case when this is useful is when the function has no side effects (other than allocation). A typical example is methods modifying collections, builders or similar objects.

Note that this is currently subject to some limitations: if the reference was passed in as an argument, returned from another function or obtained from a pointer the warning will not trigger. This may change in the future and it may start producing more warnings in some cases.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

Whenever a function paramter is marked with `#[must_use_output]` and the caller does not later access the value a warning is emitted similar to using the `+=` operator on integers and later not using the value.

> It is an error to put this attribute on parameters with types other than reference types. Pointers, smart pointers and generics are forbidden mainly for simplicity of initial implementation. They can be implemented and allowed later if it's even possible at all.

Reviewer: ```suggestion
It is an error to put this attribute on parameters with types other than reference types. Pointers, smart pointers, and generics that hide the reference, are forbidden mainly for simplicity of initial implementation. They can be implemented and allowed later if it's even possible at all.
```

Qualifying this because `&mut T` should be fine, as the generic isn't hiding the reference.

The compiler makes no distinction whether the reference is unique or shared because writes can happen through both, there doesn't seem to be a reason to forbid `Freeze` references and there is at least one case when even `Freeze` reference is useful.

The standard library functions get annotated with this attribute as well, including but not limited to:

- Modifications of collections that don't return a value
- `core::mem::swap`
- `Clone::clone`
- `*_assign` methods in `core::ops` traits
- `OpenOptions` and similar builders

# Drawbacks
[drawbacks]: #drawbacks

> - Adds another atrribute to the language complicating it

Reviewer: ```suggestion
- Adds another attribute to the language, complicating it
```

- In some cases analysis will fail and perhaps people will have false sene of security
- People could misuse it to attempt to enforce soundness of `unsafe` code (which it cannot do)

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

- We could also simply not do this but the potential mistakes it catches are real.

> - The name could be something different and preferably shorter. The name used here was suggested by Josh Triplett and is pretty readable and it can be used before stabilization in the absence of better one.

Reviewer: ```suggestion
- The name could be something different and preferably shorter. The name used here was suggested by Josh Triplett and is pretty readable and it can be used before stabilization in the absence of better one.
  - We could write `#[must_use]` on a reference parameter instead. The downside would be that this could be mistaken for saying that the *callee* must use the parameter, rather than the *caller*.
```

> - Make it a `clippy` lint instead. However not everyone uses `clippy` and the need to communicate which arguments are considered important would diminish its effect.

Reviewer: ```suggestion
- Make it a `clippy` lint instead. However not everyone uses `clippy` and the need to communicate which arguments are considered important would diminish its effect. `#[must_use]` is a native rustc lint, and this should be as well, using the same machinery.
```

- Feature Name: `must-use-output`
- Start Date: 2025-02-16
- RFC PR: [rust-lang/rfcs#3773](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

> Allow adding `#[must_use_output]` attribute on function paramters that triggers a warning if the function was called with a value that is later not used (only dropped). Also add the attribute to relevant library functions such as `Vec::push`.

Reviewer 2: ```suggestion
Allow adding a `#[must_use_output]` attribute on function parameters that triggers a warning if the function was called with a reference to a value that is later not used (only dropped). Also add the attribute to relevant library functions such as `Vec::push`.
```

# Motivation
[motivation]: #motivation

Today, if you write this code in Rust:

```rust
fn main() {
    let mut x = 0;
    x += 42;
}
```

you'll get a warning about unused write. It's likely that either you forgot to use the value of `x` or you wrote a needless operation that has no meaningful effect on the program.

However this code:

```rust
fn main() {
    let mut vec = Vec::new();
    vec.push(42);
}
```

> has a similar problem and yet it doesn't trigger a warning. This can be even dangerous in some situations. For instance, if the `Vec` is collecting multiple errors (to report messages similar to `rustc`) and the code is supposed to check if the `Vec` is empty before proceeding.

Reviewer 2: ```suggestion
has a similar problem and yet it doesn't trigger a warning. This can even be dangerous in some situations; for instance, if the `Vec` is collecting multiple errors (to report messages similar to `rustc`) and the code is supposed to check if the `Vec` is empty before proceeding.
```

Note also that some functions may be useless to call even when they take in truly immutable (`Freeze`) references. For instance `clone` is wasteful if the original is not accessed.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

You may use the `#[must_use_output]` attribute in front of any function argument with a reference type (`&T` or `&mut T`) to indicate to the caller that calling the function is useless or dangerous if the referenced value is not later accessed, e.g. reading it to check status. If the caller doesn't access the value later in the function a warning will be emitted.

As an example, this code:

```rust
fn main() {
    let mut vec = Vec::new();
    vec.push(42);
}
```

will emit a warning saying "`vec` is not used after the call to `push`" because `Vec::push` has `self` marked with this attribute:

```rust
impl<T> Vec<T> {
    pub fn push(#[must_use_output] &mut self, item: T) { /* ... */ }
}
```

This is conceptually similar to `#[must_use]` on the function, but for an output parameter rather than the return value. Like `#[must_use]`, it helps you (or downstream consumers) to catch likely mistakes.

> A common case when this is useful is when the function has no side effects (other than allocation). A typical example is methods modifying collections, builders or similar objects.

Reviewer 2: This sentence is confusing to me. It seems like you are considering mutating `*self` to not be a side effect? I don’t think that’s a default interpretation of the phrase.

(Certainly there is a *meaningful* category of “doesn't have observable effects except through `&mut` and return value”, but I don’t think there is a standard term for that yet.)

Author: I thought the word "side effect" in programming means "anything a function does that's not expressed in the arguments". I'll have to look it up.

Note that this is currently subject to some limitations: if the reference was passed in as an argument, returned from another function or obtained from a pointer the warning will not trigger. This may change in the future and it may start producing more warnings in some cases.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> Whenever a function paramter is marked with `#[must_use_output]` and the caller does not later access the value a warning is emitted similar to using the `+=` operator on integers and later not using the value.

Reviewer 2: ```suggestion
Whenever a function parameter is marked with `#[must_use_output]`, and the caller does not later access the value, a warning is emitted similar to using the `+=` operator on integers and later not using the value.
```

> It is an error to put this attribute on parameters with types other than reference types. Pointers, smart pointers, and generics that hide the reference, are forbidden mainly for simplicity of initial implementation. They can be implemented and allowed later if it's even possible at all.

Reviewer 2: ```suggestion
It is an error to put this attribute on parameters with types other than reference types. Raw pointers, smart pointers, and generics that hide the reference, are forbidden, mainly for simplicity of initial implementation. They can be implemented and allowed later, if it's even possible at all.
```

> The compiler makes no distinction whether the reference is unique or shared because writes can happen through both, there doesn't seem to be a reason to forbid `Freeze` references and there is at least one case when even `Freeze` reference is useful.

Reviewer 2: ```suggestion
The compiler makes no distinction whether the reference is unique or shared because writes can happen through both; there doesn't seem to be a reason to forbid `Freeze` references, and there is at least one case when even a `Freeze` reference is useful.
```

The standard library functions get annotated with this attribute as well, including but not limited to:

- Modifications of collections that don't return a value
- `core::mem::swap`
- `Clone::clone`
- `*_assign` methods in `core::ops` traits
- `OpenOptions` and similar builders

# Drawbacks
[drawbacks]: #drawbacks

- Adds another attribute to the language, complicating it
- In some cases analysis will fail and perhaps people will have false sene of security
- People could misuse it to attempt to enforce soundness of `unsafe` code (which it cannot do)

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

- We could also simply not do this but the potential mistakes it catches are real.

> - The name could be something different and preferably shorter. The name used here was suggested by Josh Triplett and is pretty readable and it can be used before stabilization in the absence of better one.

Reviewer 2: Possible alternate name: `#[must_read]`. This does have the "who must?" ambiguity, but I think that is not hard to learn; the name will hint it is similar to `#[must_use]` which is also an instruction to the caller.

Author: Oh, I like that!

> - The name could be something different and preferably shorter. The name used here was suggested by Josh Triplett and is pretty readable and it can be used before stabilization in the absence of better one. A pretty short and clear `#[must_read]` was also suggested by Kevin Reid.

Reviewer: For the record, while I *mildly* prefer `must_use_output`, I would happily check a box for `must_read`, and it seems like a fine choice too.

Reviewer 3: maybe `#[caller_must_use]`?

Author: @kennytm Sadly still pretty long but I like it better than `must_use_output`. It at least doesn't pose "what do you mean by the output?" question.

Reviewer: @Kixunil Agreed. `must_use` and `must_read` don't answer "who". `must_use_output` specifies that indirectly: "the one that gets this as an output must use it". But `caller_must_use` specifies it *directly*. That seems pretty clear to me.

> - We could write `#[must_use]` on a reference parameter instead. The downside would be that this could be mistaken for saying that the *callee* must use the parameter, rather than the *caller*.

Reviewer 4: This doesn't seem like much of a problem to me.  We already lint on unused function parameters (except for `self`, unfortunately, due to not having a good way to suppress it), so if I saw `#[must_use]` on a `&mut` parameter, I'd expect this would refer to an obligation on the caller, especially as that's the existing connotation of `#[must_use]`.

Barring better proposals (and hopefully more concise ones than `#[must_use_output]`), my own preference is to just use `#[must_use]` for this.

Author: Yeah, I also feel like it wouldn't be that big deal and people would figure it out quickly. But at the same time I'm not bothered by the name too much and I respect whatever decision will be made by the appropriate team.

Reviewer 5: I also prefer `#[must_use]`. When applied to a function, it also means "the output at the call site must be used", and people don't have a problem understanding it.

Reviewer 6: Maybe the attribute could be named differently, something along the lines of the function having side effect that needs to be observed `#[must_use(side_effect)]` or `#[is_a_setter]`?

Reviewer: We have a couple of parallel discussions occurring about the name, one here and one at https://github.com/rust-lang/rfcs/pull/3773#discussion_r1974960959 .

From that discussion: `caller_must_use` seems like it removes the ambiguity.

- Make it a `clippy` lint instead. However not everyone uses `clippy` and the need to communicate which arguments are considered important would diminish its effect. `#[must_use]` is a native rustc lint, and this should be as well, using the same machinery.

> - Try to somehow analyze the code and not require the attribute. This seems hard and it could lead into libraries accidentally triggering warnings in users code if they change the body of the function.
> - Implement a more full feature - e.g. by doing some things in "future posibilities" section. However this feature should be useful even without them.
> - Have the atribute on the function instead listing the names of paramters. This could make it nicer to extend to support "or" relationship described in "Future possibilities"

Reviewer 2: ```suggestion
- Try to somehow analyze the code and not require the attribute. This seems hard and it could lead into libraries accidentally triggering warnings in users’ code if they change the body of the function.
- Implement a more full feature — e.g. by doing some things in "Future possibilities". However, this feature should be useful even without them.
- Have the attribute on the function instead listing the names of parameters. This could make it nicer to extend to support the "or" relationship described in "Future possibilities".
```

> - Have the attribute on the function instead listing the names of paramters. This could make it nicer to extend to support the "or" relationship described in "Future possibilities".

Reviewer: ```suggestion
- Have the attribute on the function instead listing the names of parameters. This could make it nicer to extend to support the "or" relationship described in "Future possibilities".
```

# Prior art
[prior-art]: #prior-art

I'm not aware of anything other than the unused write warning and `#[must_use]` attribute which are somewhat related.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- Attribute name.
- Perhaps some limitations are not as difficult as I imagine and they could be lifted right away.
- Should we emit a warning if a function parameter does not have that attribute but is passed to a function with the attribute to encourage propagation?

# Future possibilities
[future-possibilities]: #future-possibilities

A custom message could be added just like `#[must_use = "..."]` is already available.

The limitations above might be lifted somehow. For instance I think that it'd be useful to also emit a warning if the reference was obtained from a function parameter that is itself *not* marked `#[must_use_output]`.

Have a way to explain to the compiler how smart pointers work so that this can be used on `Pin` as well.

> Have an attribute that can express "or" relationship between parameters and return types. For instance, `Vec::pop` is sometimes used to remove the last element and then pass the vec to something else (e.g. trim `\n` at the end) and sometimes it's used to get the last value and the vec is no longer relevant. Someting like `#[must_use_at_least_one(self, return)]` on function could handle this case.

Reviewer 5: This should just be the default behaviour when a function has multiple outputs (e.g. an output parameter as well as a return type). You have to use _at least one,_ not all of them.

Reviewer: > This should just be the default behaviour when a function has multiple outputs (e.g. an output parameter as well as a return type) that should be used. You have to use at least one, not all of them.

Then we would need a way to override it in the *other* direction, because there are absolutely functions for which you need to use all the output parameters.

Author: Yes, to give a specific example: `core::mem::swap`.

But also for `write` and `read` you should use the returned value specifically. Just using the writer is not enough.
